python-package/lightgbm/basic.py

```diff
@@ -2830,19 +2833,27 @@ def set_weight(
 
         Parameters
         ----------
-        weight : list, numpy 1-D array, pandas Series or None
+        weight : list, numpy 1-D array, pandas Series, pyarrow Array, pyarrow ChunkedArray or None
             Weight to be set for each data point. Weights should be non-negative.
 
         Returns
         -------
         self : Dataset
             Dataset with set weight.
         """
-        if weight is not None and np.all(weight == 1):
-            weight = None
+        # Check if the weight contains values other than one
+        if weight is not None:
+            if _is_pyarrow_array(weight):
+                if pa_compute.all(pa_compute.equal(weight, 1)).as_py():
```

Comment:
I'm not that familiar with `pyarrow`'s computation logic so sorry if these are ignorant questions but:

1. will this work comparison with the integer literal `1` work if the array has a `float` dtype?
2. is this expensive?
    - I don't recall exactly why the Python package has this logic to treat an array of 1s as if `weight` weren't set at all
    - probably just as an optimization to eliminate unnecessary memory allocation and computation on the training side (since a weight of `1` wouldn't actually impact training)
    - just saying I think we could choose to change the behavior if you think checking whether all values are `1` is expensive

I'm not expecting a precise answer to `2)` so don't spend too much time investigating it, just curious if it's something you thought about.

Reply:
> will this work comparison with the integer literal 1 work if the array has a float dtype?

Yes :)

<img width="456" alt="Screenshot 2023-11-08 at 10 29 15" src="https://github.com/microsoft/LightGBM/assets/22455425/fc8f8530-d629-480f-8972-35ddaa164234">

> is this expensive?

Given the following mini-benchmark, I would argue that the execution time is negligible :smile:

<img width="568" alt="Screenshot 2023-11-08 at 10 30 19" src="https://github.com/microsoft/LightGBM/assets/22455425/6a043983-e3c5-4024-b63a-8447480b3df9">

Reply:
Excellent, thank you! I thought that'd probably be the case, just wanted you to know you had options. Thanks for checking.